gemfire-core/src/main/java/com/gemstone/gemfire/cache/GemFireSparkConnectorCacheFactory.java

```diff
@@ -1,46 +1,157 @@
+    synchronized (CacheFactory.class) {
+      DistributedSystem ds = null;
+      if (this.dsProps.isEmpty()) {
+        // any ds will do
+        ds = InternalDistributedSystem.getConnectedInstance();
+      }
+      if (ds == null) {
+        ds = DistributedSystem.connect(this.dsProps);
+      }
+      PoolFactory defaultPF = null;
+       String defaultGrid =  gfeGridMappings.remove(GemFireSparkConnectorCacheImpl.gfeGridPropPrefix);
+      if (defaultGrid != null) {
+        defaultPF = this.createAndConfigurePoolFactory(defaultGrid);
+      }
+      Cache cache = GemFireSparkConnectorCacheImpl.create(defaultPF, gfeGridMappings, ds, cacheConfig);
+
+      for (Map.Entry<String, String> gridEntry : gfeGridMappings.entrySet()) {
+        String key = gridEntry.getKey().trim();
+        String gridName = key.substring(key.indexOf('.') + 1);
+        String locators = gridEntry.getValue().trim();
+        PoolFactory pf = this.createAndConfigurePoolFactory(locators);
+        pf.create(gridName);
+      }
+      return cache;
+    }
+  }
+
+
+  private PoolFactory createAndConfigurePoolFactory(String remoteLocators) {
+    PoolFactory pf = PoolManager.createFactory();
+    pf.setReadTimeout(30000);
+
+    StringTokenizer remoteLocatorsTokenizer = new StringTokenizer(remoteLocators, ",");
+    DistributionLocatorId[] locators = new DistributionLocatorId[remoteLocatorsTokenizer
+        .countTokens()];
+    int i = 0;
+    while (remoteLocatorsTokenizer.hasMoreTokens()) {
+      locators[i++] = new DistributionLocatorId(remoteLocatorsTokenizer.nextToken());
+    }
+    List<ServerLocation> servers = new ArrayList<ServerLocation>();
+    for (DistributionLocatorId locator : locators) {
+      try {
+
+        InetSocketAddress addr = new InetSocketAddress(locator.getHost(), locator.getPort());
+        GetAllServersRequest req = new GetAllServersRequest("");
+        Object res = TcpClient.requestToServer(addr.getAddress(), addr.getPort(), req, 2000);
+        if (res != null) {
+          servers.addAll((List<ServerLocation>)((GetAllServersResponse)res).getServers());
+        }
+      } catch (Exception e) {
+        System.out.println("Unable to get remote gfe servers from locator = " + locator);
+      }
+    }
+
+    List<ServerLocation> prefServers = null;
+    if (servers.size() > 0) {
+      String sparkIp = System.getenv("SPARK_LOCAL_IP");
+      String hostName = null;
+      try {
+        hostName = sparkIp != null ? InetAddress.getByName(sparkIp).getCanonicalHostName() :
+            InetAddress.getLocalHost().getCanonicalHostName();
+      } catch (Exception e) {
+        hostName = "";
+      }
+      int spacing = servers.size() / 3;
+      if (spacing != 0) {
+        prefServers = new ArrayList<ServerLocation>();
+        ServerLocation localServer = null;
+        for (int j = 0; j < servers.size(); ++j) {
+          ServerLocation sl = servers.get(j);
+          if (localServer == null && sl.getHostName().equals(hostName)) {
```

Comment:
@ahshahid
If GemFire is in a separate subnet then it would likely be configured to return public hostname in GetAllServersResponse otherwise SnappyData connector won't be able to talk to it. This will be set both on the Gem locator and well as BridgeServers and in that case the "remote-locators" option on SnappyData point to public hostname but you are right this will not help in matching against servers.

I think we can assume that if Gem settings have the hostname-for-clients then the SnappyData cluster will also have the corresponding hostname-for-clients option on the thrift servers for its public hostname and we can match the two. So then we can use Snappy's NetworkInterfaceImpl.getHostNameForClients() result instead -- will need to be made public (FabricServiceManager.currentFabricServiceInstance.getAllNetworkServers() and pick the first one checking for possible null result of currentFabricServiceInstance).